### DISCUSSION TITLE
How to work with WCF ?

### DISCUSSION THREAD

**Discussion by mitti2000** [ThreeMammals/Ocelot#1760] [Q&A] [answered]
## Expected Behavior / New Feature
I'd like to use Ocelot with an existing WCF endpoint but I can't get it to work.

## Actual Behavior / Motivation for New Feature
Following setup:
The **Route** for the WCF Service
```json
{
      "DownstreamPathTemplate": "/UpdateService",
      "DownstreamScheme": "ws",
      "DownstreamHostAndPorts": [
        { "Host": "localhost", "Port": 9000 }
      ],
      "UpstreamPathTemplate": "/UpdateService"
}
```
I tried as well with the Scheme `http` without success.

AddWebSockets is implemented.

Error message from Ocelot:
```
The connection to 'localhost' failed.
Error: ConnectionRefused (0x274d).
System.Net.Sockets.SocketException No connection could be made because the target machine actively refused it <IP-Address>
```
What else am I doing wrong?

Thanks for any help.

**Comment by AhmedAssaf** [comment]
Is there any updates?

**Comment by SViradiya-MarutiTech** [comment]
Is there any updates?

**Comment by ks1990cn** [comment]
@AhmedAssaf @SViradiya-MarutiTech  Best way to do is make synchronous call of WCF from .Net Core API and use Ocelot with this.

**Comment by raman-m** [comment]
> @AhmedAssaf commented [on May 26, 2021](#discussioncomment-7412426)
> @SViradiya-MarutiTech commented [on Aug 17, 2022](#discussioncomment-7412427)

Updates on what? Where are your advices, guys?

**Comment by raman-m** [comment]
>@ks1990cn commented [on Oct 28](#discussioncomment-7412428)

Your answer is very abstract and the author will not understand it. 
It is impossible to answer the author or recommend something, because his user case is totally wrong.

**Reply by ks1990cn** [reply]
Correct @raman-m 

So what I've answered is I've seen in my organisation with millions of users hitting WCF monolith which is wrapped by .Net Core application. That .Net Core application is called from API Gateway.

Here is document which suggest how we can call WCF from .Net Core application. Atleast user can start migration if they want keeping API gateway in front.
I understand migration to new tech also becomes tough most of time if you have huge monolith. But I think we can suggest this approach also because I've seen this as reliable way to extend to new .Net vesions.

https://stackoverflow.com/questions/63490956/how-to-consume-wcf-service-programmatic-in-net-core-3-1-version 

https://www.thecodebuzz.com/consuming-wcf-web-services-in-net-core-best-practices/

Your answer is acceptable but someone can think like this also so they can atleast start using Gateway in correct way and migrate to latest tech when they are ready.

@mitti2000 @AhmedAssaf @SViradiya-MarutiTech

**Reply by ks1990cn** [reply]
Adding to this, I've seen it doesn't create much latency so user feel a gap in server to server communications.

**Reply by raman-m** [reply]
So, the author should be an Senior Software Architect in .NET framework to integrate WCF to Ocelot.
But our team has no time for such kind on integrations. There are another popular protocols to have in Ocelot...
WCF is outdated technology, not popular now.

**Comment by raman-m** [comment, marked as answer]
> @mitti2000 [on May 4, 2020](#discussion-5787379)

Hi Thomas!
You cannot use WCF and Ocelot in general!
WCF is old technology of .NET full framework as Windows platform (before .NET Core cross-planform).
So, WCF uses Windows OS system services on transport level. WCF is not included to .NET Core.
See more:
- [Transports in Windows Communication Foundation](https://learn.microsoft.com/en-us/dotnet/framework/wcf/feature-details/transports)
- [What Is Windows Communication Foundation](https://learn.microsoft.com/en-us/dotnet/framework/wcf/whats-wcf)

There is some experiments by Microsoft in migration WCF to .NET Core:
- [CoreWCF 1.0 has been Released, WCF for .NET Core and .NET 5+](https://devblogs.microsoft.com/dotnet/corewcf-v1-released/)

But Ocelot team is not going to use these protocols.

------
> The Route for the WCF Service

You cannot route WCF traffic through Ocelot!
WCF uses different protocols on transport level. But, any RESTful and HTTP(S) traffic is routed.
Try to use `http` and/or `https` schemes only. You cannot use `ws` scheme for WCF downstream services.
Ocelot can route HTTP requests with XML-data bodies of SOAP protocol. But the problem is in headers and protocol. Some headers can be lost, not forwarded in bidirectional manner. So, even SOAP is based on HTTP, but it seems you cannot route SOAP traffic via Ocelot.
You can write some Delegating Handlers life-hacks. But that will not be easy.
**I recommend to connect clients to downstream WCF services directly!**

-----

> Error message from Ocelot:
> ```
> The connection to 'localhost' failed.
> Error: ConnectionRefused (0x274d).
> System.Net.Sockets.SocketException No connection could be made because the target machine actively refused it <IP-Address>
> ```
> What else am I doing wrong?
> 

You see this error because **you cannot connect Websocket client to WCF service!**
Stop any experiments with WCF technology (.NET Full framework 3, 4)! Ocelot is .NET Core framework product (.NET Core 3+, 5, 6, 7)

**Comment by RaynaldM** [comment]
@mitti2000 If you really want to interact with WCF, you can probably do so through DelegatingHandlers. In these, you can connect and interact with WCF.
I've never tried it, but it must be possible.

**Reply by raman-m** [reply]
Yes, in delegating handler developer can manage connection to any services using any clients. But this approach is quite hacky!
So, delegating handler can "switch off" default scheme/protocol and replace with new one in the handler.

**Comment by mitti2000** [comment]
I wrote this more then 3 years ago. This topic can be closed.

**Reply by raman-m** [reply]
😭

**Reply by ks1990cn** [reply]
Hahaha🤣.. we discussed for mankind 🤣🤣 @raman-m
